# **`Feature Engineering-6`**

`Q1. Pearson correlation coefficient is a measure of the linear relationship between two variables. Suppose
you have collected data on the amount of time students spend studying for an exam and their final exam
scores. Calculate the Pearson correlation coefficient between these two variables and interpret the result.`

In [5]:
## example dataset: using scores dataset from kaggle
import pandas as pd
df = pd.read_csv("score.csv")
df.head()

Hours  Scores
0    2.5      21
1    5.1      47
2    3.2      27
3    8.5      75
4    3.5      30

In [6]:
# Calculating the pearson correlation
df.corr(method="pearson")

Hours    Scores
Hours   1.000000  0.976191
Scores  0.976191  1.000000

The pearson correlation coeffecient between Scores and Study hours is 0.97 which indicates a strong positive correlation between the two varaibles. An increase in study hours will lead to an increase in scores

`Q2. Spearman's rank correlation is a measure of the monotonic relationship between two variables.
Suppose you have collected data on the amount of sleep individuals get each night and their overall job
satisfaction level on a scale of 1 to 10. Calculate the Spearman's rank correlation between these two
variables and interpret the result.`

In [11]:
## example dataset
import pandas as pd
df = pd.DataFrame({
    "sleep" : [8,7,6,8,9,3,4,3],
    "satisfaction" : [9,8,8,6,8,3,4,3]
})
df.head()

sleep  satisfaction
0      8             9
1      7             8
2      6             8
3      8             6
4      9             8

In [12]:
## Spearman correlation
df.corr(method="spearman")

sleep  satisfaction
sleep         1.000000      0.782745
satisfaction  0.782745      1.000000

The spearman rank correlation coeffecient between sleep hours and job satisfaction is 0.783 which indicates a strong positive correlation between the two varaibles. An increase in sleep hours will lead to an increase in job satisfaction

`Q3. Suppose you are conducting a study to examine the relationship between the number of hours of
exercise per week and body mass index (BMI) in a sample of adults. You collected data on both variables
for 50 participants. Calculate the Pearson correlation coefficient and the Spearman's rank correlation
between these two variables and compare the results.`

In [22]:
# example dataset
import pandas as pd
import numpy as np
df = pd.DataFrame({
    "exercise" : np.random.randint(1,10,50),
    "bmi" : np.random.randint(15,30,50)
})
df.head()

exercise  bmi
0         4   28
1         7   24
2         7   27
3         1   18
4         4   18

In [23]:
#pearson correlation
df.corr()

exercise       bmi
exercise  1.000000  0.082213
bmi       0.082213  1.000000

In [24]:
#spearman rank correlation
df.corr(method="spearman")

exercise       bmi
exercise  1.000000  0.076866
bmi       0.076866  1.000000

The pearson correlation coeffecient for exercise and bmi = 0.082
The spearman rank correlation coeffecient for exercise and bmi = 0.076
In this above example the pearson correlation coeffecient showed marginally higher values compared to spearman rank correlation coeffecient and since both coeffecients are positive and very close to zero we can infer that from the dataset it appears that exercise and bmi have a very weak positive relation if any at best.

`Q4. A researcher is interested in examining the relationship between the number of hours individuals
spend watching television per day and their level of physical activity. The researcher collected data on
both variables from a sample of 50 participants. Calculate the Pearson correlation coefficient between
these two variables.`

In [38]:
import pandas as pd
import numpy as np
df = pd.DataFrame({
    "tv" : np.random.randint(1,12,50),
    "activity": np.random.randint(1,4,50)
})

In [39]:
df.corr()

tv  activity
tv        1.000000 -0.079581
activity -0.079581  1.000000

In [40]:
df.corr(method="spearman")

tv  activity
tv        1.000000 -0.129281
activity -0.129281  1.000000

The pearson correlation coeffecient for v watching per day and activity = -0.079 The spearman rank correlation coeffecient for tv watching per day and activity = -0.129281 In this above example the pearson correlation coeffecient showed a higher value compared to spearman rank correlation coeffecient and since both coeffecients are negative and closer to zero we can infer that from the dataset it appears that watching tv is weakly correlates negatively with activity and as number of hours of tv watching increase, the physical activity hours are decreasing. Spearman coeffecient infers a higher negative correlation when compared to pearson coeffecient

`Q5. A survey was conducted to examine the relationship between age and preference for a particular
brand of soft drink. The survey results are shown below:`Age(Years)
25 Coke
42 Pepsi
37 Mountain dew
19 Coke
31 Pepsi
28 Coke






In [23]:
#create df
import pandas as pd
df = pd.DataFrame({
    "Age(Years)" : [25,42,37,19,31,28],
    "Soft drink Preference" : ["Coke","Pepsi","Mountain dew","Coke","Pepsi","Coke"]
})

In [24]:
#using target guided ordinal encoding
mean_age = df.groupby("Soft drink Preference")["Age(Years)"].mean().to_dict()

In [25]:
df["soft_drink_encoded"]  = df["Soft drink Preference"].map(mean_age)

In [26]:
df

Age(Years) Soft drink Preference  soft_drink_encoded
0          25                  Coke                24.0
1          42                 Pepsi                36.5
2          37          Mountain dew                37.0
3          19                  Coke                24.0
4          31                 Pepsi                36.5
5          28                  Coke                24.0

In [27]:
df[["Age(Years)","soft_drink_encoded"]].corr()

Age(Years)  soft_drink_encoded
Age(Years)             1.00000             0.83753
soft_drink_encoded     0.83753             1.00000

In [28]:
df[["Age(Years)","soft_drink_encoded"]].corr(method="spearman")

Age(Years)  soft_drink_encoded
Age(Years)            1.000000            0.833238
soft_drink_encoded    0.833238            1.000000

mean age of soft drink preference and age have a pearson correlation coeffecient of 0.837 and a spearman correlation of 0.833 both of which indicate a strong positive correlation wherein increase in age will also result in an change(increase) in the mean age of soft drink preference (from coke to mountain dew)

`Q6. A company is interested in examining the relationship between the number of sales calls made per day
and the number of sales made per week. The company collected data on both variables from a sample of
30 sales representatives. Calculate the Pearson correlation coefficient between these two variables.`

In [50]:
#random df
import pandas as pd
import numpy as np
df = pd.DataFrame({
    "sales_calls_per_day" : np.random.randint(1,51,30),
    "sales_per_week": np.random.randint(11,61,30)
})

In [51]:
df.corr() #pearson correlation

sales_calls_per_day  sales_per_week
sales_calls_per_day             1.000000        0.340796
sales_per_week                  0.340796        1.000000

The pearson correlation for sales calls per day and number of sales per week is 0.340796 which indicates a weak positive correlation